src/uopool/services/simulation.rs

```diff
@@ -0,0 +1,131 @@
+use ethers::{
+    abi::AbiDecode,
+    contract::ContractError,
+    prelude::HttpClientError,
+    providers::{Http, Provider, ProviderError},
+};
+
+use crate::{
+    contracts::entry_point::{self, SimulationResult},
+    types::user_operation::UserOperation,
+    uopool::services::uopool::UoPoolService,
+};
+
+#[derive(Debug)]
+pub enum BadUserOperationError {
+    EntryPointSimulateValidation { result: String },
+}
+
+#[derive(Debug)]
+pub enum UserOperationSimulationError {
+    Simulation(BadUserOperationError),
+    Internal(anyhow::Error),
+    Provider(ProviderError),
+}
+
+impl From<anyhow::Error> for UserOperationSimulationError {
+    fn from(e: anyhow::Error) -> Self {
+        UserOperationSimulationError::Internal(e)
+    }
+}
+
+impl From<ProviderError> for UserOperationSimulationError {
+    fn from(e: ProviderError) -> Self {
+        UserOperationSimulationError::Provider(e)
+    }
+}
+
+impl UoPoolService {
+    async fn entry_point_simulate_validation(
+        &self,
+        user_operation: &UserOperation,
+    ) -> Result<SimulationResult, UserOperationSimulationError> {
+        // View call to simulateValidation(userop)
+        let simulation_response = self
+            .entry_point
+            .simulate_validation(entry_point::UserOperation::from(user_operation.clone()))
+            .await;
+
+        let result = format!("{:?}", simulation_response);
+
+        if let Err(ContractError::MiddlewareError(ProviderError::JsonRpcClientError(
+            json_rpc_client_error,
+        ))) = simulation_response
```

Comment:
I am also working on the API and I found the error handling is very bad in ethers-rs(I asked the ethers-rs team and they agreed :) ). This error could be deeply embedded if people choose different embedded middleware. I chose a different way to handle errors here. 
I just put a comment here to mark. We could revisit the difference later.

Reply:
Yeah, I agree, it's very bad.

It's great that you chose a different way; we can compare and see what's better. Calls to the entry-point smart contract should probably be moved to a separate file.